### ISSUE TITLE
Android API level needs to be visible in Swift

### ISSUE THREAD

**Issue by grynspan** [swiftlang/swift#76671]
### Motivation

In Android, new API is guarded by API level. For instance, `timespec_get()` was introduced with Android API level 29. API level can be determined by checking the value of `__ANDROID_API__` defined in the system header `<android/api-level.h>`.

This value is _not_ visible to Swift, meaning that code that is conditionally dependent on the availability of some API needs to check for it (and provide a fallback path!) in C or C++ rather than in Swift.

### Proposed solution

I propose we teach the compiler, when targetting Android, to expose the API level similar to how the Swift language mode and compiler version are exposed:

```swift
var ts = timespec()
#if _apiLevel(>=29)
timespec_get(&ts, TIME_UTC)
#else
clock_gettime(CLOCK_REALTIME, &ts)
#endif
return ts
```

### Alternatives considered

- Exposing the API level as a macro would result in it being evaluated too late and the compiler wouldn't be able to ignore the API usage.
- Exposing the API level as the _de facto_ OS version (`#available(Android 29, *)`) would have a similar problem.
- "Just do it in C" is generally _not_ the solution we want for Swift projects.

**Comment by grynspan** [comment]
@finagolfin @compnerd Could use your thoughts here. :)

**Comment by grynspan** [comment]
> There's another option of mapping OS release to API level, but that isn't very precise either.

See my "alternatives" section, yeah. `#available` would potentially be evaluated too late to be useful unless Android has weak linking support à la Darwin.

**Comment by bc-lee** [comment]
> For Android, something like that with a weak linkage in the indeterminate case would be really nice.

It is possible. The macro `__ANDROID_UNAVAILABLE_SYMBOLS_ARE_WEAK__` is specifically designed for this purpose. When defined, the NDK treats all symbols unavailable on the current platform as weak, similar to how it's done on Darwin.

For more information, see [this guide](https://developer.android.com/ndk/guides/using-newer-apis#other_build_systems) and refer to the implementation details in [android/versioning.h](https://cs.android.com/android/platform/superproject/main/+/main:bionic/libc/include/android/versioning.h;drc=ed2d3832bd8afef5021c3efc6db3352f2b35a848;l=20) and [time.h](https://cs.android.com/android/platform/superproject/main/+/main:bionic/libc/include/time.h;drc=55acfc1a1cf13a4ef9d8f0e290a839575d101414;l=462).

**Comment by bc-lee** [comment]
I think so. https://github.com/llvm/llvm-project/blob/5ef02a3fd4758ae1b9151ac581eebd1109b4daad/compiler-rt/lib/builtins/os_version_check.c#L321

**Comment by finagolfin** [comment]
It's something Android porters like @drodriguez have talked about for years, but nobody got around to adding. Since Android usually builds with the API version in the build triple, it should be easy to pass in at compile time to Swift code, just as it is currently passed in by SwiftPM when compiling C/C++ code with clang and Swift's `ClangImporter`, which then [invoke this code that sets it in clang](https://github.com/llvm/llvm-project/blob/fb78495376017111d8b5a1f43745d3c17b0d8978/clang/lib/Basic/Targets/OSTargets.h#L326). So only Swift code is currently oblivious of the Android API used, past time that was remedied.

**Comment by finagolfin** [comment]
We should assign somebody to work on this: @compnerd, is this something you Browser Company people would add? It all depends on who needs it, whether @marcprux for Skip or you guys or somebody else entirely.

**Comment by compnerd** [comment]
This is definitely something that would be nice to have. This is similar to the long standing feature request on Windows for availability support. I think that would allow us to surface both of these in the best way possible, but that isn't something that we can take on just right now with the other items that are outstanding.

**Comment by grynspan** [comment]
I think we would also want the PSG to chime in if we were ready to pursue a change here for any given platform.

**Comment by marcprux** [comment]
> [@marcprux](https://github.com/marcprux), given your Android toolchain, is this something you Skip people would want to add?

While I agree it would be very useful, we unfortunately don't have the bandwidth to tackle it anytime in the near future.

**Comment by finagolfin** [comment]
Ok, for me personally, this is a nice to have feature but not a priority, so I'm sounding the community out to see who is willing to put the time in to implement and push this through. Just for completeness, pinging @ephemer and @mstokercricut to see if either of you have any input.

**Comment by ephemer** [comment]
Thanks for the ping. We would also like to see this, but it's not something we actively have a current use-case for.

**Comment by al45tair** [comment]
Yes, I think this is potentially something PSG is interested in. _Personally_ I'd rather see `@/#available` used for this if at all possible, instead of adding a whole new thing, and then we could use the Android version number as part of the target perhaps, in a similar way to the way Darwin systems do?

**Comment by finagolfin** [comment]
Alright, looks like this is just waiting on someone to pitch and implement it then.

**Comment by madsodgaard** [comment]
I'd like to tackle this. Although, I was thinking of just adding a compile-time directive such as the `_apiLevel` originally suggested, at least for now. So no runtime conditionals. We have a use case for this when we need to change our Swift code depending on conditionals in the Bionic library such as the one below, which changes the signature available to Swift from Bionic.

```
#if __ANDROID_API__ < 24
#include <bits/struct_file.h>
#endif
```

What do you think? @compnerd @al45tair @grynspan

**Comment by compnerd** [comment]
> Yes, I think this is potentially something PSG is interested in. _Personally_ I'd rather see `@/#available` used for this if at all possible, instead of adding a whole new thing, and then we could use the Android version number as part of the target perhaps, in a similar way to the way Darwin systems do?

I agree, that this should be handled as `@available` and `#available`. Fortunately, turns out that the compiler-rt support for this is already present in the form of `__isOSVersionAtLeast`. We already use the Android API level in the target triple (at least once it is normalised). So, this should fit pretty much exactly into the Darwin version handling mold.

> Although, I was thinking of just adding a compile-time directive such as the _apiLevel originally suggested, at least for now.

I don't think that we should introduce `_apiLevel` as we already have the necessary support in place, it is now wiring it up to the frontend, package manager, build systems.

**Comment by grynspan** [comment]
We may _also_ need `_apiLevel()` for static availability checks because Swift may be compiling against an older NDK that doesn't have a newer symbol (specifically `timespec_get()` is a motivating example here).

But I'll take what I can get TBH.

**Comment by madsodgaard** [comment]
@compnerd But #available only gives me runtime checks. If I for example want to define one typealias for API level < 23 and another for 23+, because the NDK has different signatures. I would need `#if _apiLevel`. Is that not right?

**Comment by finagolfin** [comment]
> Swift may be compiling against an older NDK that doesn't have a newer symbol

I don't think we should support older NDKs. The latest NDKs all support back to Android API 21 with updated headers, I think we should only support LTS NDK 27 and later.

> If I for example want to define one typealias for API level < 23 and another for 23+, because the NDK has different signatures. I would need #if _apiLevel. Is that not right?

How does that work on Darwin platforms now? To avoid such compile-time issues, I would imagine @/#available also has some way to do that, but I've never built Swift for Darwin, so never used that.

**Comment by bc-lee** [comment]
I think unifying conditional availability syntax matters.

For existing Apple platform Swift code, this pattern is very common:

```swift
if #available(iOS 13, macOS 10.15, tvOS 13, watchOS 6, *) {
    // Use modern API
} else {
    // Fallback on earlier versions
}
```

On Android in C++, this is also possible:

```cpp
if (__builtin_available(android 29, *)) {
  // Introduced in API level 29
  AHardwareBuffer_lockPlanes(buffer, AHARDWAREBUFFER_USAGE_CPU_READ_OFTEN, -1,
                             nullptr, planes);
} else {
  // Fallback on earlier versions
}
```

If I remember correctly, both languages use the same underlying compiler_rt builtin, `__builtin_available`.

So, even for the Android case, I think it would be better to use `#available` syntax to unify with Swift style, for example:

```swift
func lockHardwareBuffer29(buffer: OpaquePointer, usage: Int, fence: Int,
                          rect: UnsafePointer<Rect>?,
                          planes: UnsafeMutablePointer<Plane?>) -> Int32 {
    if #available(android 29, *) {
        return AHardwareBuffer_lockPlanes(buffer, usage, fence, rect, planes)
    } else {
        // Do something else
        ...
    }
}
```
